swarm/src/stream.rs

```diff
@@ -1,16 +1,43 @@
 use futures::{AsyncRead, AsyncWrite};
 use libp2p_core::muxing::SubstreamBox;
 use libp2p_core::Negotiated;
-use std::io::{IoSlice, IoSliceMut};
-use std::pin::Pin;
-use std::task::{Context, Poll};
+use std::{
+    io::{IoSlice, IoSliceMut},
+    pin::Pin,
+    sync::{Arc, Weak},
+    task::{Context, Poll},
+};
 
 #[derive(Debug)]
-pub struct Stream(Negotiated<SubstreamBox>);
+pub struct Stream {
+    stream: Negotiated<SubstreamBox>,
+    counter: StreamCounter,
+}
+
+#[derive(Debug)]
+enum StreamCounter {
+    Arc(Arc<()>),
+    Weak(Weak<()>),
+}
 
 impl Stream {
-    pub(crate) fn new(stream: Negotiated<SubstreamBox>) -> Self {
-        Self(stream)
+    pub(crate) fn new(stream: Negotiated<SubstreamBox>, counter: Arc<()>) -> Self {
+        let counter = StreamCounter::Arc(counter);
+        Self { stream, counter }
+    }
+
+    /// Opt-out this stream from the [Swarm](crate::Swarm)s connection keep alive algorithm.
+    ///
+    /// By default, any active stream keeps a connection alive. For most protocols,
+    /// this is a good default as it ensures that the protocol is completed before
+    /// a connection is shut down.
+    /// Some protocols like libp2p's [ping](https://github.com/libp2p/specs/blob/master/ping/ping.md)
+    /// for example never complete and are of an auxiliary nature.
+    /// These protocols should opt-out of the keep alive algorithm using this method.
+    pub fn no_keep_alive(&mut self) {
```

Comment:
```suggestion
    pub fn ignore_for_keep_alive(&mut self) {
```

Thinking more about this, I think this is a more descriptive name. What do you think?

Reply:
Agree with this, it would be straightforward.